### ISSUE TITLE
Add flat maps for all fsaverage resolutions

### ISSUE THREAD

**Issue by alexisthual** [nilearn/nilearn#3171]
It feels to me that plotting flat maps is a common thing to do in neuroscience, yet we don't really advertise this in `nilearn` even though it's already possible.

I think we could:
1. include flat maps for both hemispheres for all `fsaverage` resolutions.
We would compute them with [`mris_flatten`](https://surfer.nmr.mgh.harvard.edu/fswiki/mris_flatten) and convert outputs to gifti files
1. add an example using these surfaces

Also, currently, one has to tweak the `view` kwarg to fit the orientation of the flattened surface.
Maybe we could make sure that our maps display nicely when `view` is `lateral`, but this is a minor comment IMHO.

### Benefits to the change

Much plotting, better beautiful, fosterz package swag :tada: 

![Screenshot from 2022-03-02 17-29-41](https://user-images.githubusercontent.com/13835654/156405307-6fc1a875-08b7-49f6-a29e-6635b5190d6b.png)

![Screenshot from 2022-03-02 17-33-34](https://user-images.githubusercontent.com/13835654/156405545-434e3f20-2705-4fdf-a13f-5fa1bc259bd1.png)

Images above were generated with this snippet (I found the [`fsaverage7` flattened gifti files](https://figshare.com/articles/dataset/fsaverage_subject_for_pycortex/9916166) online):

```python
import matplotlib.pyplot as plt
import nibabel as nib
from nilearn import datasets, plotting
import numpy as np

# %%
mesh_path = "/home/alexis/singbrain/fsaverage_flat/fsaverage/surfaces/flat_lh.gii"
mesh = nib.load(mesh_path)
n_voxels = mesh.darrays[0].data.shape[0]

surf_map = np.random.rand(n_voxels)

fs7 = datasets.fetch_surf_fsaverage(mesh="fsaverage7")

# %%
plotting.plot_surf(
    mesh_path,
    surf_map,
    bg_map=fs7.sulc_left,
    bg_on_data=True,
    view="dorsal"
)
plt.show()

# %%
fig = plotting.plot_surf(
    mesh_path,
    surf_map,
    bg_map=fs7.sulc_left,
    bg_on_data=True,
    view="dorsal",
    engine="plotly"
)
_ = fig.show(renderer="nteract")
```

**Comment by bthirion** [comment]
+1 for adding flat maps for plotting.
My impression was that those from pycortex were nicer (for fsaverage). Shouldn't we use those instead ?
Bzw when you plot a flat map, you really need to add a texture depicting sulco-gyral anatomy.

**Comment by alexisthual** [comment]
> My impression was that those from pycortex were nicer (for fsaverage). Shouldn't we use those instead ?

I think the one plotted in the examples above comes from pycortex.

> Bzw when you plot a flat map, you really need to add a texture depicting sulco-gyral anatomy.

Yes, that's a must-do!

**Comment by emdupre** [comment]
Were you able to find flat maps at other resolutions, @alexisthual  ? I'm not sure if @mvdoc would have recommendations on where to access these !

**Comment by mvdoc** [comment]
The fsaverage surface flatmap we use in pycortex was manually generated by @marklescroart, and is publicly available on [figshare](https://figshare.com/articles/dataset/fsaverage_subject_for_pycortex/).

I tried automating the generation of fsaverage{5,6} flatmaps from what Mark did, but it doesn't work really well (or at least, all my attempts failed). The best option would be to manually reflatten all other fsaverage resolutions, while maintaining the same cuts (so that the flatmap will look approximately the same across resolutions).

Another alternative (which I use most of the times when plotting flatmaps of fsaverage{5,6}) is projecting the same data (nearest neighbor projection) from the vertices of fsaverage{5,6} to fsaverage7. This is obviously for visualization only, and it will add some smoothness. In pycortex we implement a method to recover and cache the surf2surf freesurfer matrices to map from fsaverage{5,6} to fsaverage7 ([cortex.freesurfer.get_mri_surf2surf_matrix](https://github.com/gallantlab/pycortex/blob/68692281d8a68c31045077939f8b3da748946320/cortex/freesurfer.py#L638)). With that it's fairly straightforward to plot fsaverage{5,6} data to fsaverage7 (`data @ projection_matrix`). There are probably more principled ways to obtain the neighborhood of vertices between fsaverage7 and fsaverage{5,6} by exploiting how the reduced resolution templates are generated, but this works well and it's quick.

**Comment by alexisthual** [comment]
Thank you for the figshare link @mvdoc (I think it's broken though, but it's most likely the one I used to generate the previous plots).

I think there is a way to generate flat meshes for fsverage{5,6} using that of fsaverage7:
1. determine which vertices of fsaverage7 are also vertices of fsaverage5 (I just checked, using vertices' spherical coordinates, that fs5 vertices are a subset of fs7's)
2. compute fs7's adjacency matrices `a1` and `a2` before and after it was flattened (we actually need powers of these matrices to a high enough rank d ; I assume d=10 should be enough)
3. remove faces of fs5 containing an edge (i, j) such that `a1[i,j] != a2[i,j]` (if the shortest path from i to j is longer after flattening, it means intermediate edges between these two voxels have been removed, and we should remove them from fs5's flattened mesh)
4. assign fs5 vertices with coordinates of matching vertices in fs7

Does anyone else think this would work? :slightly_smiling_face:

**Comment by mvdoc** [comment]
Whoops sorry, this is the correct link to figshare: https://figshare.com/articles/dataset/fsaverage_subject_for_pycortex/9916166

The approach you describe is similar to what I tried. However, the resulting surfaces looked like crap near the cuts. Flattening a surface results in distortions for the vertices/edges close to cuts, and I think that for this reason you can't simply decimate a flattened surface (as you can with an intact surface). But I tried some time ago and there's always the possibility that I did something wrong ;)

**Comment by alexisthual** [comment]
Sooo, I tried the method I described above. In my humble opinion, results are good enough that we should integrate them in `nilearn`. What do you think @bthirion and @jeromedockes? (anyone else is welcome to join the discussion of course)

Here are flat map plots for fsaverage 3 to 7:

![image](https://user-images.githubusercontent.com/13835654/202770322-46b94ceb-4a5d-4f20-a53d-28b01c00d639.png)

![image](https://user-images.githubusercontent.com/13835654/202770304-7a66aae9-6f3a-4e1b-8844-f5234f18c622.png)

![image](https://user-images.githubusercontent.com/13835654/202770292-d55842a8-eba5-4c66-8700-bdfd6fd4c46c.png)

![image](https://user-images.githubusercontent.com/13835654/202770262-6183c580-8c4f-4ddd-942a-92a9a460e46d.png)

![image](https://user-images.githubusercontent.com/13835654/202770232-254078df-f5b2-4058-a832-92cb9de79722.png)

In short, for a given fsaverage `a` mesh, I keep faces for which all associated edges `(i, j)` have not been affected in flattening fsaverage `b` (where `b` > `a`). My criterion for deciding whether an edge has been affected or not is: has the shortest path between i and j changed before and after flattening `b`? If yes, then it's likely that this edge should be deleted in `a`.
I had to reorder fsaverage 3 and 4's vertices accordingly to what is explained in #3415.
